### PR TITLE
Raft - do not retry claims for held leases

### PR DESCRIPTION
**Pull request by manadart**
When there is churn around Raft in a HA scenario - as for HA establishment, leader elections or controller restarts, there can be a situation where the lease manager checks the local store before issuing a claim and finds no lease. If the local store is backed by a Raft FSM that is not in sync with the leader, the lease can actually be held by another entity, which is reported back as an invalid lease claim resulting in up to 10 retries.

Example of abridged logs:
```
machine-1: 11:46:06 TRACE juju.worker.lease.raft [3d712b] machine-1 asked for lease 3d712b1b-1835-475f-8ba4-675cb536ce57, no lease found, claiming for 1m0s
machine-1: 11:46:08 TRACE juju.core.raftlease got response, err invalid lease operation
machine-1: 11:46:08 ERROR juju.core.raftlease command Command(ver: 1, op: claim, ns: singular-controller, model: fb97b3, lease: 3d712b1b-1835-475f-8ba4-675cb536ce57, holder: machine-1): invalid lease operation
machine-1: 11:46:08 TRACE juju.core.raftlease runOnLeader claim, elapsed from publish: 1.118s
machine-1: 11:46:08 INFO juju.worker.lease.raft [3d712b] got invalid lease operation after 10 retries, denying claim {"singular-controller" "fb97b3ec-9ff1-490b-8fdd-8f8b61641fc2" "3d712b1b-1835-475f-8ba4-675cb536ce57"} for "machine-1"
```

This patch adds a new error type that is returned by the FSM and handled by the manager so that the claim is rejected without retries, lessening the load on the manager, pub-sub hub and lease store.

## QA steps

- `juju bootstrap serverstack-manadart serverstack-test --no-gui --debug --build-agent --config use-floating-ip=true --config network=manadart_admin_net --config logging-config="<root>=DEBUG;juju.worker.lease.raft=TRACE;juju.core.raftlease=TRACE"`
- Set up a workload to have some lease claims kicking around.
- `juju deploy postgresql -n 2`
- `juju deploy telegraf`
- `juju relate telegraf postgresql:juju-info`
- `juju deploy content-cache`
- `juju deploy nrpe`
- `juju relate telegraf content-cache`
- `juju relate nrpe content-cache`
- `juju deploy ubuntu`
- `juju relate telegraf ubuntu`
- `juju deploy nagios`
- `juju relate nrpe nagios`
- `juju relate telegraf nagios:juju-info`
- Starting watch the logs: `juju debug-log -m controller --include-module juju.core.raftlease --include-module juju.worker.lease.raft --include-module juju.worker.lease.raft`
- One the workloads quiesc, in another terminal, enable HA: `juju enable-ha`
- Claims for leases held by other entities should now be logged like this:
```
machine-6: 13:28:25 TRACE juju.worker.lease.raft [be92a3] machine-6 asked for lease be92a334-c7fb-416e-8ddd-3fb124004a84, no lease found, claiming for 1m0s
machine-6: 13:28:25 TRACE juju.core.raftlease got response, err lease already held
machine-6: 13:28:25 ERROR juju.core.raftlease command Command(ver: 1, op: claim, ns: singular-controller, model: 03e134, lease: be92a334-c7fb-416e-8ddd-3fb124004a84, holder: machine-6): lease already held
machine-6: 13:28:25 TRACE juju.core.raftlease runOnLeader claim, elapsed from publish: 82ms
machine-6: 13:28:25 TRACE juju.worker.lease.raft [be92a3] machine-6 asked for lease be92a334-c7fb-416e-8ddd-3fb124004a84, held by by another entity; local Raft node may be syncing
```

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Raft - do not retry claims for held leases](https://trello.com/c/jbJpAztc/3081-raft-do-not-retry-claims-for-held-leases)

File: core/lease/errors.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 Canonical Ltd.
```
Comment (manadart) [inline core/lease/errors.go:1]: These are just relocated to their own file from separate areas of the code base.

File: core/lease/errors.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+	// ErrInvalid indicates that a Store operation failed because latest state
+	// indicates that it's a logical impossibility. It's a short-range signal to
+	// calling code only; that code should never pass it on, but should inspect
+	// the Store's updated Leases() and either attempt a new operation or return
+	// a new error at a suitable level of abstraction.
+	ErrInvalid = errors.New("invalid lease operation")
+
+	// ErrHeld indicates that a claim operation was impossible to fulfill
+	// because the has been claimed on behalf of another entity.
```
Comment (wallyworld) [inline core/lease/errors.go:27]: typo

**Review by wallyworld — APPROVED** [review]
Nice

**Comment by wallyworld** [comment]
$$merge$$

**Comment by manadart** [comment]
@wallyworld I found an issue with this. The denied claim immediately adds a block, which is immediately released (lease not in the local store), which in turn causes another claim. This cycles quite fast. I found it when enabling HA using this patch.

**Pull request merged**